insights/parsers/audit_log.py

```diff
@@ -0,0 +1,164 @@
+            * **unparsed** - the unparsed part in log, may or maynot exist
+
+            If the line is not regular, for some reason, then as many fields as
+            possible are pulled from the line.
+        """
+        info = {}
+        linesp = shlex.split(line)
+
+        if (len(linesp) < 2 or
+                not (linesp[0] and linesp[0].startswith('type=')) or
+                not (linesp[1] and linesp[1].startswith('msg=audit('))):
+            return
+
+        info['type'] = linesp[0].lstrip('type=')
+
+        timestamp_id = linesp[1].lstrip('msg=audit(').rstrip('):')
+        timestamp_id_sp = timestamp_id.split(':')
+        if len(timestamp_id_sp) != 2:
+            return
+        info['timestamp'] = timestamp_id_sp[0]
+        info['msg_ID'] = timestamp_id_sp[1]
+
+        for index in xrange(len(linesp) - 1, 1, -1):
+            itemsp = linesp[index].split('=', 1)
+            if len(itemsp) < 2:
+                unparsed1 = (line.rsplit(linesp[index + 1])[0]
+                             if index + 1 < len(linesp)
+                             else line)
+                unparsed2 = unparsed1.split(linesp[1])[-1]
+                info["unparsed"] = unparsed2.strip()
+                break
+            info[itemsp[0]] = itemsp[1]
+
+        info['raw_message'] = line
+        return info
+
+    def get(self, s):
+        """
+        Parameters:
+            s(str or list): one or more strings to search for.
+
+        Returns:
+            (list): list of dictionaries corresponding to the parsed lines
+                contain the `s`.
+
+        Examples::
+
+            [
+             {'timestamp': '1506487181.009',
+              'msg_ID': '32794',
+              'type': 'AVC',
+              'unparsed': 'avc:  denied  { create } for',
+              'raw_message': '...: ...'
+              ...
+             }, ...
+            ]
+        """
+        r = []
+        for l in self.lines:
+            if ((type(s) == list and all(w in l for w in s)) or
+                    (type(s) == str and s in l)):
+                info = self._parse_line(l)
+                if info:
+                    r.append(info)
+        return r
+
+    def get_after(self, timestamp, s=None):
+        """
+        Find all the (available) logs that are after the given time stamp.
+        Override this function in class LogFileOutput.
+
+        Parameters:
+
+            timestamp(datetime.datetime): lines before this time are ignored.
+            s(str or list): one or more strings to search for.
+                If not supplied, all available lines are searched.
+
+        Yields:
+            (dict): the parsed data of lines with timestamps after this date in the
+            same format they were supplied.
+        """
+        for line in self.lines:
+            # If `s` is not None, keywords must be found in the line
+            if (s and ((type(s) == str and s not in line) or
```

Comment:
1) I believe it's best practice to compare types with 'isinstance' rather than directly.
2) I'd recommend for speed to set a comparison lambda function outside the loop and then use it inside the loop, e.g.:
```python
if not s:
    return []
if isinstance(s, str):
    comp=lambda line: s in line
elif isinstance(s, list):
    comp=lambda line: any(w in line for w in s)
else:
    raise TypeError('search items must be given as list or string')
for line in self.lines:
    if not comp(line):
        continue
    ...
```

Reply:
In general it might be a good idea to think about how we can abstract this 'search from list or string' behaviour to all 'get_after' methods.

Reply:
If you mean the 'search for list of strings or string' behaviour, this is already apllied to class `LogFileOutput`  in https://github.com/RedHatInsights/insights-core/pull/686 . 

And I'd like to take the 2 suggestions above. Thanks.